src/security_utilities_rust/src/microsoft_security_utilities_core/identifiable_secrets.rs

```diff
@@ -811,3 +814,111 @@ fn index_of(input_string: String, search_char: char) -> i32
+        let mut current_detection : Option<Detection> = None;
+
+        detections.sort_unstable_by_key(|item| item.start());
+
+        for detection in detections.iter() {
+            let scan_match = detection.matches_bytes(input_as_bytes, true).unwrap();
+            let match_text = scan_match.text();
+
+            if validate_checksum {
+                let match_text_as_bytes = match_text.as_bytes();
+                let mut signature_bytes = vec![0; 3];
+                signature_bytes[0] = match_text_as_bytes[57];
+                signature_bytes[1] = match_text_as_bytes[58];
+                signature_bytes[2] = match_text_as_bytes[59];
+
+                let signature = general_purpose::STANDARD.encode(&signature_bytes);
+
+                let checksum_validation_result = try_validate_common_annotated_key(
+                    &match_text,
+                    &signature,
+                );
+
+                assert!(checksum_validation_result);
+            }
+
+            // Get c3id for the find
+            let match_text_c3id = cross_company_correlating_id::generate_cross_company_correlating_id(match_text);
+
+            let redaction_token = match default_redaction_token {
+                Some(token) => token,
+                None => &format!("SEC101/200:{}", match_text_c3id),
+            };
+
+            if current_detection.is_none() {
+                current_detection = Some(Detection {
+                    start: detection.start(),
+                    end: detection.start() + detection.len() as u64,
+                    length: detection.len(),
+                    redaction_token: redaction_token.to_string()
+                });
+                current_detections.push(current_detection.clone().unwrap());
+            } else {
+                if detection.start() <= current_detection.clone().unwrap().end {
+                    // Overlapping case or contiguous case.
+                    let current = current_detection.as_mut().unwrap();
+                    current.length = (std::cmp::max(current.end,
+                                                    detection.start() + detection.len() as u64)
+                        - current.start) as usize;
+                } else {
+                    current_detection = Some(Detection {
+                        start: detection.start(),
+                        end: detection.start() + detection.len() as u64,
+                        length: detection.len(),
+                        redaction_token: redaction_token.to_string()
+                    });
+                    current_detections.push(current_detection.clone().unwrap());
+                }
+            }
+        }
+
+        let mut index_adjustment: usize = 0;
+        for detection in current_detections {
+            let start_index = detection.start as usize - index_adjustment;
+            input.replace_range(start_index..start_index + detection.length, &detection.redaction_token);
```

Comment:
@beaubelgrave I am piggybacking on `replace_range` to do the in-place replacement, followed by some math adjustment on the indices. Is this a good choice?

Reply:
I think so.

Since you modify the input now, I wouldn't return an Option<String>. Maybe just return a bool indicating if the text was modified/changed?

If I were to read this function signature I would have thought the Option<String> would be returning something like the modified string, but I don't think that's the intention anymore.